main.go

```diff
@@ -1,196 +1,861 @@
+			} else {
+				goType = *v
+			}
+		}
+
+		isNullable = isNullable || fd.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE || fd.GetType() == descriptor.FieldDescriptorProto_TYPE_GROUP
+		if v, ok := fd.OptionExtensions["gogoproto.nullable"].(*bool); ok {
+			isNullable = *v
+		}
+
+		goName := generator.CamelCase(fd.GetName())
+		if v, ok := fd.OptionExtensions["gogoproto.customname"].(*string); ok {
+			goName = *v
+		}
+		if v, ok := fd.OptionExtensions["gogoproto.embed"].(*bool); ok && *v {
+			goName = goType
+		}
+
+		if fd.OneofIndex != nil {
+			oneOfType := fmt.Sprintf("%s_%s", fm.GetName(), goName)
+			if fm.GetMessage(goName) != nil {
+				oneOfType = fmt.Sprintf("%s_", oneOfType)
+			}
+			oneOfName := generator.CamelCase(fm.GetOneofDecl()[fd.GetOneofIndex()].GetName())
+
+			srcPath = fmt.Sprintf("%s.Get%s()", srcPath, goName)
+			dstPath = fmt.Sprintf("%s.%s", dstPath, oneOfName)
+
+			fmt.Fprintf(buf, `
+			switch {
+			case %s != nil && nilPath:
+			case %s == nil && nilPath:
+				continue
+			case %s == nil:
+				%s = &%s{}
+			}
+`,
+				dstPath,
+				dstPath,
+				dstPath,
+				dstPath, oneOfType,
+			)
+
+			dstPath = fmt.Sprintf("%s.(*%s).%s", dstPath, oneOfType, goName)
+
+		} else {
+			srcPath = fmt.Sprintf("%s.%s", srcPath, goName)
+			dstPath = fmt.Sprintf("%s.%s", dstPath, goName)
+		}
+
+		isRepeated := fd.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED
+		if isNullable || isRepeated {
+			fmt.Fprintf(buf, `
+			if nilPath || %s == nil {
+				%s = nil
+				continue
+			}`,
+				srcPath,
+				dstPath,
+			)
+		} else {
+			fmt.Fprintf(buf, `
+			if nilPath {
+				var v %s
+				%s = v
+				continue
+			}`,
+				goType,
+				dstPath,
+			)
 		}
+
+		if isRepeated {
+			if strings.HasSuffix(fd.GetTypeName(), "Entry") && mdMap[fd.GetFullName()] == nil {
+				// fd is a map field, however obtaining its type is non-trivial, hence we resort to reflection
+				fmt.Fprintf(buf, `
+			%s`, deepCopyOp(dstPath, srcPath))
+				continue
+			}
+
+			if isNullable {
+				goType = fmt.Sprintf("[]*%s", goType)
+			} else {
+				goType = fmt.Sprintf("[]%s", goType)
+			}
+
+			var copyStr string
+			for _, l := range strings.Split(copyOp(fmt.Sprintf("%s[%s]", dstPath, "i"), "v"), "\n") {
+				copyStr += fmt.Sprintf(`
+				%s`,
+					l,
+				)
+			}
+
+			fmt.Fprintf(buf, `
+			%s = make(%s, len(%s))
+			for i, v := range %s {%s
+			}`,
+				dstPath, goType, srcPath,
+				srcPath, copyStr,
+			)
+
+			continue
+		}
+
+		if isNullable {
+			fmt.Fprintf(buf, `
+			var v %s
+			%s = &v`,
+				goType,
+				dstPath,
+			)
+
+			srcPath = fmt.Sprintf("(*%s)", srcPath)
+			dstPath = fmt.Sprintf("(*%s)", dstPath)
+		}
+
+		var copyStr string
+		for _, l := range strings.Split(copyOp(dstPath, srcPath), "\n") {
+			copyStr += fmt.Sprintf(`
+			%s`,
+				l,
+			)
+		}
+		fmt.Fprint(buf, copyStr)
 	}
+	imports["fmt"] = "fmt"
+	fmt.Fprintf(buf, `
+		default:
+			panic(fmt.Errorf("Invaild fieldpath: '%%s'", path))
```

Comment:
Not sure if `panic` is a good idea during setting the fields. Though the fieldmasks will be validated in the API boundaries and it's unlikely that an invalid field will get through that,  I think an `interal error` with cause should be sufficient.

Reply:
@htdvisser ?